ui/src/notebooks/context/utils.ts

```diff
@@ -0,0 +1,119 @@
+// Libraries
+import {isEqual, uniq} from 'lodash'
+
+export const dedupeTags = (tags: any[]) => {
+  const cache = {}
+  const set = new Set()
+  let results = []
+  tags.forEach(tag => {
+    Object.entries(tag).forEach(([tagName, tagValues]) => {
+      // sorting tagValues to check for exact matches when doing string comparisons
+      const values = tagValues as string[]
+      values.sort()
+      const vals = JSON.stringify(values)
+      if (tagName in cache === false) {
+        cache[tagName] = values
+        set.add(vals)
+        results = results.concat({
+          [tagName]: values,
+        })
+        return
+      }
+      if (isEqual(cache[tagName], values) || set.has(vals)) {
+        return
+      } else {
+        set.add(vals)
+        results = results.concat({
+          [tagName]: values,
+        })
+      }
+    })
+  })
+  return results
+}
+
+const filterFields = (
+  fields: string[],
+  searchTerm: string,
+  selectedField: string
+): string[] => {
+  return fields.filter((field: string) => {
+    if (!!selectedField) {
+      return field === selectedField
+    }
+    return field.includes(searchTerm)
+  })
+}
+
+const filterTags = (tags: any[], searchTerm: string): any[] =>
+  tags.filter(
+    tag =>
+      Object.entries(tag).filter(([tagName, tagValues]) => {
+        const values = tagValues as any[]
+        if (tagName.includes(searchTerm)) {
+          return true
+        }
+        return values?.some(val => val.includes(searchTerm))
+      }).length !== 0
+  )
+
+export const normalizeSchema = (schema: any, data: any, searchTerm: string) => {
```

Comment:
seems like data is pretty well defined? maybe it should have a type

Reply:
Most def